File: 4.DSA/algorithms/java/blind_75/arrays_and_hashing/TwoSum.java
import java.util.HashMap;
import java.util.Map;

public class TwoSum {

    public int[] twoSum(int[] nums, int target) {
        Map<Integer, Integer> map = new HashMap<>();
        for (int i = 0; i < nums.length; i++) {
            int diff = target - nums[i];
            if (map.containsKey(diff)) {
                int[] result = new int[2];
                result[0] = map.get(diff);
                result[1] = i;
                return result;
            }
            map.put(nums[i], i);
        }
        return new int[2];
    }
}

File: 5.Rust/dsa/src/twosum.rs
use std::collections::HashMap;

pub fn two_sum(nums: Vec<i32>, target: i32) -> Vec<i32> {
    let mut map: HashMap<i32, i32> = HashMap::new();

    for (index, &value) in nums.iter().enumerate() {
        if let Some(&diff) = map.get(&(target - value)) {
            return vec![index as i32, diff];
        }
        map.insert(value, index as i32);
    }

    vec![]
}

#[cfg(test)]
mod tests {
    use super::*;
    
    #[test]
    fn test_two_sum() {
        let nums = vec![2, 7, 11, 15];
        let target = 9;
        let result = two_sum(nums, target);
        assert_eq!(result, vec![1, 0]);
    }
}
